### PR TITLE
Use static_check util

### LINKED ISSUE #23: Use `static_check` util everywhere
All tests have static checks. Let's use the newly added `static_check` util everywhere we can to make those checks more readable.

### PR DESCRIPTION
**Pull request by godexsoft**
Fixes #23 

All tests are now using static_check util. 
Not 100% sure about removing the few places we had clang-format disabled. It's not too bad as it is now but arguably was a little nicer before.

### CODE REVIEW

File: tests/tests/or_else.cpp [hunk 1]
```diff
@@ -437,27 +415,38 @@ template <typename T> constexpr auto fn_int_lvalue = [](int &) -> T { throw 0; }
...
 } // namespace
 
-// clang-format off
 static_assert(invocable_or_else<decltype(fn_Error<std::expected<Value, Error>>), std::expected<Value, Error>>);
 static_assert(invocable_or_else<decltype(fn_generic<std::expected<int, int>>), std::expected<int, int>>);
-static_assert(invocable_or_else<decltype(fn_Error<std::expected<Value, Xerror>>), std::expected<Value, Error>>); // error type conversion
-static_assert(invocable_or_else<decltype(fn_Error<std::expected<Value, Error>>), std::expected<Value, Xerror>>); // error type conversion
-static_assert(invocable_or_else<decltype(fn_Error<std::expected<Value, int>>), std::expected<Value, Xerror>>); // error type conversion
-static_assert(not invocable_or_else<decltype(fn_Error<std::expected<Value, int>>), std::expected<Value, int>>); // wrong error_type
+static_assert(invocable_or_else<decltype(fn_Error<std::expected<Value, Xerror>>),
```
Comment (godexsoft) [inline tests/tests/or_else.cpp:420]: Hm. Might roll these back. On one hand i dislike manual overrides, on the other hand this is way less readable :rage2:

**Comment by Bronek** [comment]
@godexsoft do you mind keeping the `is_p` and `is_e` tests in `filter.cpp` separate ? They do not work the same; error generator is a sink, but predicate is not (it can only read a value) and keeping these mixed creates confusing impression that they work the same. While doing so you might add more tests to `is_p` to see it in practice (no matter the type of operand, predicate won't bind a non-const lvalue-ref or rvalue-ref, but error generator will)

**Comment by Bronek** [comment]
Also, the "`// lvalue operand`" comments are misleading when the operand type has no `&` qualifier. These are prvalues, not lvalues. In case of binding to predicates parameter, we turn these to const lvalue (by means of `std::as_const`) , but predicate in `filter` (and also `inspect`) are unusual, since they are not sinks. Normal monadic operations are sinks and they would receive parameter as-is (which is what we use `FWD(v)` for, or someone else would write a little slower to compile and longer to type `std::forward<decltype(v)>(v)`)

**Comment by godexsoft** [comment]
> Also, the "`// lvalue operand`" comments are misleading when the operand type has no `&` qualifier. These are prvalues, not lvalues. In case of binding to predicates parameter, we turn these to const lvalue (by means of `std::as_const`) , but predicate in `filter` (and also `inspect`) are unusual, since they are not sinks. Normal monadic operations are sinks and they would receive parameter as-is (which is what we use `FWD(v)` for, or someone else would write a little slower to compile and longer to type `std::forward<decltype(v)>(v)`)

That's a good observation. I will update accordingly.

File: tests/util/static_check.hpp [hunk 2]
```diff
@@ -11,20 +11,86 @@
...
+                  OperationType, typename Categories<OperandType>::type, decltype(fn), HandlerTypes...>
+              && ...);
+    }
+
+    static constexpr auto invocable_with_any(auto &&fn) noexcept -> bool
+    {
+      return invocable<lvalue, cvalue, rvalue, clvalue, crvalue, prvalue>(FWD(fn));
+    }
+
+    static constexpr auto not_invocable_with_any(auto &&fn) noexcept -> bool
```
Comment (godexsoft) [inline tests/util/static_check.hpp:60]: Names for these should really be `not_invocable_with_any_value_category` but it's so long that it multilines lots of places and defeats the purpose of having this "alias" in the first place. I could not find a shorter name that still sort of made sense in this context. Please do suggest better name if you can think of one :rage3:

File: tests/util/static_check.hpp [hunk 3]
```diff
@@ -11,20 +11,86 @@
...
+
+template <typename OperandType> struct cvalue {
+  using type = std::add_const_t<OperandType>;
+};
+
+template <typename OperandType> struct prvalue {
+  using type = OperandType;
+};
+
+template <typename OperationType, typename OperandType> struct static_check {
```
Comment (godexsoft) [inline tests/util/static_check.hpp:38]: Please ignore all the copypasta - i can't be bothered.. this is just for tests anyways.

**Review by Bronek — APPROVED** [review]

**Pull request merged**